Improve: variety of MXP changes

eowmob opened the pull request:
#### Brief overview of PR changes/additions
Added our mud: dlgConnectionProfiles.cpp, aldebaran.png, mudlet.qrc
Made \b a NOP rather than printing illegal char: TBuffer.cpp
Working B,U,I,S and aliases: TBuffer.cpp, TMxpClient.h, TMxpFormattingTagsHandler.cpp, TMxpMudlet.cpp, TMxpMudlet.h
Make !EN work: TMxpEntityTagHandler.cpp
Working !EN DELETE: TEntityResolver.cpp
Don't force send hints all upper: TMxpCustomElementTagHandler.cpp
Support for ATT='TAG=default' in !EL :TMxpCustomElementTagHandler.cpp, TMxpElementDefinitionHandler.cpp, TMxpElementRegistry.h
Can now set !EN to empty string: TMxpEntityTagHandler.cpp
Made fgColors, bgColour private (because it is a right thing(TM) in C++): TMxpMudlet.h
Allow entity including '|' in the args for SEND: TMxpSendTagHandler.cpp
Allow to set a 'headline'-hint in SEND-menus: TMxpSendTagHandler.cpp, TTextEdit.cpp
Updated SUPPORT to new extensions: TMxpTagProcessor.cpp
VERSION supports a style parameter: TMxpVersionTagHandler.cpp, TMxpMudlet.h
#### Motivation for adding to Mudlet
Make it conform better to the MXPv1 protocol, s.t. it can be used to play our mud in MXP mode 
#### Other info (issues closed, discussion etc)
Forum discussion in: Bugs in MXP implementation, is this work in progress?

eowmob committed:
Basic MXP Enhancements

Added our mud: dlgConnectionProfiles.cpp, aldebaran.png, mudlet.qrc

Made \b a NOP rather than printing illegal char: TBuffer.cpp

Working <B>,<U>,<I>,<S> and aliases: TBuffer.cpp, TMxpClient.h, TMxpFormattingTagsHandler.cpp, TMxpMudlet.cpp, TMxpMudlet.h

Make <!EN> work: TMxpEntityTagHandler.cpp

Working <!EN DELETE>: TEntityResolver.cpp

Don't force send hints all upper: TMxpCustomElementTagHandler.cpp

Support for ATT='TAG=default' in !EL :TMxpCustomElementTagHandler.cpp, TMxpElementDefinitionHandler.cpp, TMxpElementRegistry.h

Can now set !EN to empty string: TMxpEntityTagHandler.cpp

Made fgColors, bgColour privat (because it is a right thing(TM) in C++): TMxpMudlet.h

Allow entity includeing '|' in the args for <SEND>: TMxpSendTagHandler.cpp

Allow to set a 'headline'-hint in <SEND>-menus: TMxpSendTagHandler.cpp, TTextEdit.cpp

Updated <SUPPORT> to new extensions: TMxpTagProcessor.cpp

<VERSION> supports a style parameter: TMxpVersionTagHandler.cpp, TMxpMudlet.h

eowmob committed:
Update TMxpSendTagHandler.cpp

Tooltip for <send>menu based on first hint, not first command.

vadi2 commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -859,6 +860,17 @@ QString dlgConnectionProfiles::getDescription(const QString& hostUrl, const quin
         return QStringLiteral(
                 "The original vision of FieryMUD was to create a challanging MUD for advanced players. This new reborne Fiery is a hope to bring back the goals of the past by inflicting certain death on unsuspecting players. FieryMUD will continue to grow and change through the coming years and those players who seek challenge and possess imagination will come in search of what the 3D world fails ...
+    } else if (hostUrl == QStringLiteral("mud.tap.de")) {
+        return QStringLiteral(
```
Thanks for the addition! Could you confirm that the game meets the requirements set out in https://wiki.mudlet.org/w/Listing_Your_MUD and add this as a separate, scoped PR

eowmob replied:
I wasn't aware of these requirements, sorry for that. The map requirements will need some mud-internal work and review with other archs, so for now, no requirements are not met. I'll single this out and get back to you if and when they are met.

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -706,6 +706,14 @@ void TBuffer::translateToPlainText(std::string& incoming, const bool isFromServe
 COMMIT_LINE:
+        // some rudimentary handling of backspace:
+        if (ch == '\b') {
+            ++localBufferPosition;
```
This isn't related to MXP and could mess with how Mudlet works for other games. Could you submit this as a separate PR, so we can test it on its own?

eowmob replied:
Ok, yes this also and indeed of minor importance in this contect. I'll single this out. I'm new to GitHub, so I might need some time seeing how to do it (probably I need to make a different branch).

For the remaining comments, they seem legit too.. I'll work this out and make another Pull request.. I think that's the right procedure (again, I'm new to githubt).

Kebap replied:
> I'm new to GitHub, so I might need some time seeing how to do it (probably I need to make a different branch).

Well, welcome to the party! 😄 Yes, you can make lots of branches, they are easy to manage. Do one for each PR. Make many small PRs, they are easier to review. Also, this may help as well: https://github.com/Mudlet/Mudlet/blob/development/.github/CONTRIBUTING.md

vadi2 commented on src/TMxpElementRegistry.h:
```diff
@@ -33,6 +33,7 @@ struct TMxpElement
     QString name;
     QString definition;
     QStringList attrs;
+    QMap<QString, QString> defval;
```
What does defval mean?

eowmob replied:
These are the default values of attributes. Consider &lt;!EL USER FORE=&f; ATT='f=red'&gt; if &lt;USER&gt; is called w/o an 'f' attribute, f will default to red. In this case we would have defval["f"]="red" .

Do you have a suggestion for a better name? Should I add a comment to the source? Thx.

vadi2 replied:
Ah - `defaultValue` is good then

vadi2 commented on src/TMxpMudlet.h:
```diff
@@ -81,11 +86,18 @@ class TMxpMudlet : public TMxpClient
     void stopMedia(TMediaData& mediaData) override;
 
-    bool isBold, isItalic, isUnderline;
+    void setBold(bool bold) override {boldCtr += bold ? 1 : -1; if(boldCtr <0) boldCtr = 0; }
```
Pull this out into the .cpp please as not a one liner

vadi2 commented on src/TMxpSendTagHandler.cpp:
```diff
@@ -32,30 +33,46 @@ TMxpTagHandlerResult TMxpSendTagHandler::handleStartTag(TMxpContext& ctx, TMxpCl
+    // no | chars in the entity names (allowed, at least), so interpolate
+    // first:
+    href = ctx.getEntityResolver().interpolate(href);
+    if (!hint.isEmpty())
```
Here and anywhere else, the consistent code style in Mudlet is braces around ifs

vadi2 reviewed (commented):
Thanks a lot for this! Appreciate the improvements to Mudlet.

I've done just a preliminary review. @gcms, as the original author would you be able to pitch in?

eowmob committed:
RevertToSplitInBranches

Reverting to create new branches for Mud submission and backspace ignore.

eowmob committed:
RevertToSplitInBranches

Reverting to create new branches for Mud submission and backspace ignore.

eowmob committed:
RevertToSplitInBranches

Reverting to create new branches for Mud submission and backspace ignore.

eowmob committed:
CleanUpMxpSendTag

Resolve conflict from Revert since a change was added later.

eowmob committed:
Added fixes from 1st Review

Working B,U,I,S and aliases: TBuffer.cpp, TMxpClient.h, TMxpFormattingTagsHandler.cpp, TMxpMudlet.cpp, TMxpMudlet.h
Make !EN work: TMxpEntityTagHandler.cpp
Working !EN DELETE: TEntityResolver.cpp
Don't force send hints all upper: TMxpCustomElementTagHandler.cpp
Support for ATT='TAG=default' in !EL :TMxpCustomElementTagHandler.cpp, TMxpElementDefinitionHandler.cpp, TMxpElementRegistry.h
Can now set !EN to empty string: TMxpEntityTagHandler.cpp
Made fgColors, bgColour private (because it is a right thing(TM) in C++): TMxpMudlet.h
Allow entity including '|' in the args for SEND: TMxpSendTagHandler.cpp
Allow to set a 'headline'-hint in SEND-menus: TMxpSendTagHandler.cpp, TTextEdit.cpp
Updated SUPPORT to new extensions: TMxpTagProcessor.cpp
VERSION supports a style parameter: TMxpVersionTagHandler.cpp, TMxpMudlet.h

eowmob committed:
Typos in Comments

fixing bad English

eowmob commented:
I think I made all the changes you asked for. Please check again.
As I'm unable to create a new pull request, I assume I'm expected to continue on this one..

eowmob commented:
And travis fails with an internal error when building its build environment, which is imho not related to my patches. I don't know how to trigger another build attempt

eowmob committed:
Extend a comment.

Clarified a comment, but main intend was to trigger another build attempt by Travis.

eowmob commented:
Tricked Travis to rebuild with a pointless comment change.

vadi2 commented on src/TMxpCustomElementTagHandler.cpp:
```diff
@@ -89,11 +89,22 @@ MxpStartTag TMxpCustomElementTagHandler::resolveElementDefinition(const TMxpElem
+ * And only when <send> is used in a custom element, but not when used directly?
+ * Let the mud choose about its user interface.
+ */
+#if 0
```
Are you sure it's forced and hasn't been converted to uppercase elsewhere?

eowmob replied:
Yes, 100%:
When using mudlet to connect to our mud I was surprised / wondering why all MXP menus appear in uppercase, though they are specified with mixed case in the mud and display as such in all other mud clients.
The patch above definitely changes this to behave 'normal'.
IMHO, it's fine when a mud wants to display MXP hints in all upper as a 'design' decision, it can just specify the hints in uppercase. But - I think the mud client should not enforce an all upper display on its own for stylistic reasons - Or if this is really something players want to have, it should be an option in the game profile - like a tick box: 'display MXP send menus in all uppercase'.
I was even more surprised, that the uppercase conversion only occurs for send-hints when send is used as a part of a custom element tag. It does not occur when send is used directly. I didn't notice this fine distinction at first because we use custom elements all over. But when I saw it in the source, I could prove it easily with an mxp testtool.

vadi2 replied:
Yeah I agree with you, and I think we haven't noticed it because it happens in a specific circumstances as you say. Not intentional. I think you can take the comment out

eowmob replied:
Ok, removed the comment, github tests ran fine.

vadi2 replied:
Excellent. Speaking of the tests, can you add ones to cover your improvements? MXP is hairy and adjusting it for one game can easily break another, so we're starting to lock it down with tests.

eowmob replied:
Yes, definitely. But I'm not sure what exactly you have in mind. Ideally, there should be some test file(s) with
different MXP scenarios (text files, basically) and then one should check if mudlet does not crash or lockup when
rendering them and more difficult, that the result looks right. Even more difficult, some things depend on a user
action (hover over a link, select a menu). Imho that's difficult to automate, not impossible (maybe for me w/o a proper tool). Is there something in place for mudlet already, which I'm not aware off?

Probably it'll be easier if one could load a file into mudlet and see how it displays and compare with an example document (which might include screen shots; - even from other clients). I assume you might be able to do this with a lua script, but it won't test the complete input chain. It would be better to connect to a mudstyle server thatjust produces the test pattern. There is this mudlet test server. I assume it does something like this, but I can't connect to. I checked my IPv6 connection with other sites, it worked, but I can't reach the test server.

I'd be willing to provide some test patterns/files, surely - I have one for my own experiments already and can enhance them. Even for things that don't (yet) work with mudlet, even lock it up (esp when a mud misbehaves). But I'd need to know exactly what you need, a text file, a doc file with example screen shots, etc..

If there's a hint in the mudlet docs or web site regarding this, I didn't find it, so just give a pointer.

Michael.

P.S.: But, there are countless combinations on using MXP, there still might be some mud using a specific, untested combination one didn't think of. But if there's a first testing suite, one could add those issues to the suite and then they won't duplicate.

vadi2 replied:
That sounds great! There's a few ways of doing them, we've went along the C++/API level. You can find the tests in https://github.com/Mudlet/Mudlet/tree/development/test and you can see them running for your PR in [here](https://github.com/Mudlet/Mudlet/pull/4200/checks?check_run_id=1329194065):

![image](https://user-images.githubusercontent.com/110988/97748370-49a03080-1ae5-11eb-926d-459aeb18bed4.png)

eowmob replied:
Sorry for the late reply, I was busy in RL. I need a helping hand, as I don't get QT Creator (on Win10) to even try compiling the test directory. What I found in the manuals didn't work out, though I tried a bunch of things, but it doesn't want to compile/perform any tests.

Regarding the tests themselves.. I'm well, uncertain on my next steps. I wrote already I consider making sensible tests really difficult. 

These tests.. I'm really speechless... well they are bare unit tests, they can test some internal functions and that's it. I fear they are actually counterproductive as they might make people think MXP is working and well tested. I was quite shocked when I looked into the Mudlet src and see that most of the MXP functionality missing from the released program is already there, or at least already laid out in the design. However, the functions are not called right by the code around them or don't have any effect beyond parsing their arguments. I can now clearly see how this situation developed. 

What I don't see is how a unit level test will help to check for or mitigate these problems. They'd help if I made up some complex internal memory structure and want some automated test calls to run and check if my memory handling is right. But checking for proper interaction with TTextedit?

I need to be able to compile the test directory, then I can see if I can make up something. However, I fear whatever I do will hinder someone else from improving my code and mudlet in general, when he - for example - optimizes some internals and made them work differently. This is explicictly not what I want to do.

vadi2 replied:
Yeah, happy to lend a hand in getting it to run the tests. Jump in on Discord.

Mudlet by design implements a small subset, not a full version, of MXP. Very few games use the full blown MXP implementations and since the standard is a good mess, we don't deal with it besides the bare minimum. That's why MXP was rewritten, only a small amount - what we actually supported before - was actually plugged in, and the rest was left available, which is what you came across.

eowmob replied:
Is it really required to use Discord? It can't be any more than a few check marks or options to make, which I just can't figure out w/o reading the whole lot of Qt docs and not wanting to risk breaking the projects definition files. I'm not into socialising by PC and I don't feel like getting a mic/headset and install discord just for a simple bugfix.

Kebap replied:
Oh. We actually use discord mostly for quick text chat much like irc but with inline screenshots, etc.  However hardly use the voice chat you assume. Also it runs in a browser, no need to install anything.

vadi2 replied:
Yeah it's our replacement nforbchat really. That said I'll look into documenting how to run the tests as well.

SlySven replied:
Yeah, for the record we used [Gitter](https://gitter.im/Mudlet/Mudlet) a couple of years back but now the only activity you'll see there is notifications from the CI builds - otherwise it is all tumbleweed there now.

You can use this permanent [invite](https://discord.com/invite/kuYvMQ9) from Vadim to access our Discord guild from a web-browser even without the client app (or if you use an OS that is not supported by the application - like the FreeBSD that I sometimes use)...

eowmob replied:
>  That said I'll look into documenting how to run the tests as well.

Yep, that would be most helpful and highly appreciated. The docs on how to rebuild mudlet were really helpful, btw. There was one or the other gap to my level of Github and QT knowledge, but it could be bridged. From what I read and understood, I just need a different project / mudlet.po file, or clone the existing and add one or more build parameters which I'm unable to figure out. I already enabled "running all tests with every build" in QTCreator, but it just does not seem to have any effect.

vadi2 replied:
I think you need a cmake build for the tests and the instructions currently use qmake. Are you on Windows?

eowmob replied:
Yes, I'm on windows (If anyone had told me I build something under Windows.. I'm really a Unix/Linux guy at least I was in the ancient days when I had time, but Windows is what I use for daily work now) . And yes, qmake or actually: C:\Qt\5.13.2\mingw73_32\bin\qmake.exe and/or C:\Qt\Tools\mingw730_32\bin\mingw32-make.exe are used. I can change the latter make executable to C:\Program Files (x86)\CMake\bin\cmake.exe in the project, and cmake is used. However, this cmake does not accept the -j<proc> parameter. I can change the <proc> parameter, but I cannot remove it.
Also, trying to do things on cli, it seems cmake want to make a build, not an executable. So probably I need to replace the qmake be cmake, but I cannot change that in the project settings.
The same directory has also a ctest, but it complains there is no test configuration file found. QT Creator, BTW, has a bunch of options running tests, but it seems, no test config is found. Hence, QT Creator does not run any tests.

eowmob replied:
Well, I managed to get it to compile with qmake and mingw32-make as installed according to your 'Compiling on Windows 7+' instructions. Unfortunately I had to learn that qmake does not really support building multiple targets. So I had to play around with creating a bunch of makefiles on the fly, etc. However with a project file (disguised with an uncommon name s.t. QtCreator), a plain Makefile to make the makefiles and a batch file to set the paths as needed (the Qt installation didn't extend the system paths.. maybe a good thing (maybe not)). Basically you 'RunTestsOnWindows' in the test directory and thats it. 

Hmm.. I can't attach the files to this post, so I'll include them to the pull request, for you to have a look.

You should think about and give me an indication, if you want my enhanced MXP tests as part of this pull request, or yet another one (IMHO, another one makes little sense, as my tests, when sensible, will definitely fail w/o my patches)

I will now:

a) Fix the commented out TMxpSendTagHandlerTest.cpp s.t. it works again.

b) Try to add some more, sensible tests. I have some ideas to make them at least somewhat sensible. However, as I already tried to elaborate: There was (almost) no issue with parsing MXP, hence I didn't change more than a few tiny bits of the parser, the issues were in how MXP functions were called from the rest of the code (or weren't) and how they interacted with TBuffer and friends. IMHO, any at least barely sensible test would include some kind of 'fake' TBuffer which will record the outputs of the MXP stages send to the display in some kind of text form which can then checked for. IMHO, a sensible test of MXP (or any other aspect of the terminal emulation) would include some 'mud' input, ideally coming in through all layer of the tcp and telnet handling and then checking the visual output of the mudlet window... where actually, OS and version specific variations regarding fonts, exact placement of tooltips.. exact 'RGB' values of standard colours... and and and are OS and implementation specific, hence slight variations are totally fine.

Well...
Enjoy.

eowmob replied:
Hi Vadim,
We have some serious OS/ISP Problems at our game that will require my undivided attention / free time for several days, maybe weeks. Due to that I cannot currently enhance the test code. I'll get back checking for this, but I cannot yet foresee when.

eowmob committed:
Update TMxpCustomElementTagHandler.cpp

Removed comment/preprocessor option regarding send.hint transformed to all upper.

eowmob committed:
Batch and Makefiles to run tests on windows

Scripts and Makefiles to perform tests on Windows.

eowmob committed:
Rebased on development 10.1

Resolved conflicts to get a compiling version

eowmob committed:
Further conflict resolution

Resolved further conflicts from the rebase to 4.10 which did not yield compile time errors, change typos / clarified some comments

eowmob committed:
fixed existing MXP tests

Remove some unused parser variables
Fixed / Added Windows Makefile to compile TMxpSendTagHandlerTest
Fixed test to deal with send menu hints
Made MxpStubClient stack MXP char attributes

eowmob committed:
DefaultAttribute case-insensitive, more MXP module tests.

Fixed/Extended and wrote new MXP module tests to refer to bugs fixed in this brunch.
Fixed inappropriate case-sensitivity when handling default Attributes

eowmob committed:
CMakeList for Tests patched/White space changes in test dir

Fixed one missing source file in CMakeList (could not test on Windows)
Tabs changed to spaces in Test Sources
Added std Copyright notice to Test Sources

eowmob committed:
Update CMakeLists.txt in test

Include path to edbee-lib

eowmob committed:
Update CMakeLists.txt

Corrected path the edbee-lib, 2nd attempt

eowmob committed:
Update CMakeLists.txt

Try removing TMxpMusicTagHandler from Test to avoid lua path issues

eowmob committed:
Update CMakeLists.txt

Remove SoundTagHandler from CMake file in tests to avoid compilation errors

eowmob committed:
Update CMakeLists.txt

Trying to find path to required Lua library

eowmob committed:
Update CMakeLists.txt

Still trying to find lua5.1 library.. 3rd attempt

eowmob committed:
Update CMakeLists.txt

Giving up, commented out Test which don't compile since the installed lua library is not found.

eowmob committed:
Fixed Compiler Warnings in TMxpStubClient.h

added missing overrides

vadi2 committed:
Revert dblsqd submodule change

vadi2 committed:
Revert edbee submodule change

vadi2 committed:
Reset vcpkg submodule

vadi2 committed:
Merge remote-tracking branch 'origin/development' into MXP-enhancements

vadi2 commented:
This is good for review again @SlySven @gcms

keneanung commented on test/CMakeLists.txt:
```diff
@@ -29,9 +29,10 @@ link_libraries(
 
-
-include_directories("${CMAKE_CURRENT_SOURCE_DIR}/../src")
+include_directories("${CMAKE_CURRENT_SOURCE_DIR}/../src" "${CMAKE_CURRENT_SOURCE_DIR}/../3rdparty/edbee-lib/edbee-lib" "/usr/include/lua51")
```
1. Hardcoding the `edbee-lib` include path here is wrong. See https://github.com/Mudlet/Mudlet/blob/development/src/CMakeLists.txt#L398 for the correct way to include those headers.
2. If you use `find_package` for lua (which is the right thing to do), hardcoding the include path here is wrong. See https://github.com/Mudlet/Mudlet/blob/development/src/CMakeLists.txt#L401 how the include should be handled.

eowmob replied:
>Very late to the party
Better late than never, as a few versions have been released since then, I thought the patch would just be dropped as MXP to the extend I'd like to use is probably no prio.
I'll try your hints. All these are only for auto TEST being run. I've not been able to find the proper paths in the build system. Vadi wasn't able to hint me.  I assume I'll have to rejoin with the current devel path since some time passed.  I'll get back to you regarding this.

eowmob replied:
Hello keneanung,

I had no luck rebuilding on Github, C-Style checks now bail out in countless Mudlet and internal QT library files I never touched. I have not the slightest clue how to fix this.

I'm outta here. Submitting these IMHO relatively simple patches to Mudlet needs much too much effort for me. At least twice as much of the time than I can afford. I saw countless issues with MXP that need to be fixed to comply with standards and this was only meant as a first test patch to make it barely work, and I'm already unable to keep track with build system changes and stuff.

I give up, sorry. I understood Mudlet's idea of what mudclient is and should be differs from mine and has thus different priorities - which is totally ok - , but I'm outta here. Sorry.. I'm done, I can't help it.

Good luck with your project.

vadi2 replied:
@eowmob Sorry about that. Part of the reason nobody had the willpower to pick this up was because it was way too many changes in one PR (we don't allow this anymore) and in an area that is arguably difficult to work in.

That said we've dusted off your PR now and will be looking at incorporating your changes in a future Mudlet update.

keneanung commented on test/Makefile.TestsOnWindows:
```diff
@@ -0,0 +1,24 @@
+all: Makefile_TEntityHandlerTest Makefile_TEntityResolverTest Makefile_TLinkStoreTest Makefile_TMxpTagParserTest Makefile_TMxpSendTagHandlerTest Makefile_TMxpEntityTagTest Makefile_TMxpVersionTagTest Makefile_TMxpCustomElementTagHandlerTest
```
This seems to be autogenerated and should not be checked in.

eowmob replied:
No, it is not. There is no Cmake in the Mudlet Build System you proposed on you wiki and I have to use the Windows system. This Makefile may not be the best one ever written, but it allowed me to compile the test code on Windows which I wasn't able to be otherwise. If you don't want it in the official tree.. fine for me. but other Windows builders won't be able to run the test code with Cmake then (no cmake in QtCreator)

vadi2 replied:
Qt Creator now supports CMake https://doc.qt.io/qtcreator/creator-project-cmake.html 👍

keneanung commented on test/MudletTest.qmake:
```diff
@@ -0,0 +1,83 @@
+######################################################################
```
This seems to be autogenerated and should not be checked in.

eowmob replied:
Maybe, will doublecheck.. Obviously this comment line is of no importance.

keneanung commented on test/RunTestsOnWindows.cmd:
```diff
@@ -0,0 +1,4 @@
+echo If it doesn't work, check if the version numbers in the paths are correct!
```
I'm unsure about this file, to be honest... It seems unintended to me, but I could be wrong.

eowmob replied:
As said above, there is no Cmake in QTCreator on Windows. And QtCreator does also not build these test snippets. This batch file runs qmake to perform the tests w/o me entering all path names on the command line. I can remove it, but it was meant as a helping hand for someone wanting to build an run the test snippets on Windows.  I have my .. well reservations .. about this test code anyway, because it only tests isolated library functions whereas most of the issues I found are in how things are working together in the global Mudlet framework. But Vadi insisted on the tests. To me, they are fun but only of little use to check the overall functioning of MXP within mudlet.

I can remove these helping scripts to compile the tests on Windows, no problem. But someother Windows builder might find them helpful.

keneanung reviewed (dismissed):
Very late to the party, I know :frowning: But there are some things I'm not quite agreeing with.

eowmob committed:
Update CMakeLists.txt

keneanung committed:
Merge branch 'development' into MXP-enhancements

keneanung committed:
Fix linker error when linking TMxpSendTagHandlerTest

keneanung committed:
Fix TMxpTagParserTest::testSimpleElementDefinition

keneanung committed:
Remove unneeded files

keneanung committed:
Fix comment about VERSION tag returning stuff

vadi2 commented on src/TTextEdit.cpp:
```diff
@@ -1016,7 +1016,12 @@ void TTextEdit::updateTextCursor(const QMouseEvent* event, int lineIndex, int tC
+                if (tooltip.size() > commands.size())
+                    QToolTip::showText(event->globalPos(), tooltip[0]);
+                else
+                    QToolTip::showText(event->globalPos(), tooltip.join("\n"));
```
```suggestion
                if (tooltip.size() > commands.size()) {
                    QToolTip::showText(event->globalPos(), tooltip[0]);
                } else {
                    QToolTip::showText(event->globalPos(), tooltip.join("\n"));
                }
```

vadi2 committed:
Update src/TTextEdit.cpp

keneanung committed:
Allow tests for custom elements to run

SlySven commented on src/TBuffer.cpp:
```diff
@@ -824,12 +824,12 @@ void TBuffer::translateToPlainText(std::string& incoming, const bool isFromServe
+                | (mStrikeOut || mpHost->mMxpClient.isStrikeOut() ? TChar::StrikeOut : TChar::None)
+                | (mUnderline || mpHost->mMxpClient.isUnderline() ? TChar::Underline : TChar::None);
 
         TChar c((!mIsDefaultColor && mBold) ? mForeGroundColorLight : mForeGroundColor, mBackGroundColor, attributeFlags);
```
In passing I am not sure if the logic on selecting between the first or second set of 8 colours in the 16-colour set for the foreground also needs tweaking here as well as on line 827...

:man_shrugging:

SlySven commented on src/TEntityResolver.cpp:
```diff
@@ -51,7 +51,7 @@ bool TEntityResolver::registerEntity(const QString& entity, const QString& str)
 }
 
 bool TEntityResolver::unregisterEntity(const QString & entity){
```
```suggestion
bool TEntityResolver::unregisterEntity(const QString & entity)
{
```

SlySven commented on src/TMxpMudlet.h:
```diff
@@ -36,11 +36,18 @@ class TMediaData;
+
+    // These are also a kind of stack for parameters as fg/bgColours, but here a
+    // simple counter suffices:
+    int boldCtr, italicCtr, underlineCtr, strikeOutCtr;
```
Why are we not initialising them to `0` here and doing it in the constructor?

vadi2 replied:
No reason, we can do it. Want to submit a change?

SlySven replied:
You are already (re)working on it - so me also adding to it is going to get confusing - do I add anything as a PR onto @eowmob 's `MXP-enhancements` or onto something else...?

The main reason I picked up on this was actually the definition of multiple members of the same type on a single line without initialisation - that always makes me itch (or more so if only the last one being initialised!)

vadi2 replied:
Submit a one-line suggestion instead of asking why are we doing it this way, it does not seem like a useful question to ask.

vadi2 committed:
Update src/TEntityResolver.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on test/TEntityHandlerTest.cpp:
```diff
@@ -1,3 +1,24 @@
+/***************************************************************************
+ *   Copyright (C) 2020 by Gustavo Sousa - gustavocms@gmail.com            *
+ *   Copyright (C) 2020 by Stephen Lyons - slysven@virginmedia.com         *
```
I don't get this - why is **my** name being dragged into **this** and some other files that I don't think I ever had any involvement with?

vadi2 replied:
Probably because you insisted on adding the GPLv2 header earlier so they just blank-added you. @SlySven the person who worked on these changes is not present in the PR anymore for a year, there's no use leaving comments for them.

keneanung committed:
Replace Stephen's name in copyright header

Since he has not touched the files, he did not want his name among the
copyright holders yet. Replaced them with mine as I definitely did an
edit now.

keneanung committed:
Fix codefactor issues

keneanung committed:
Apply fixes suggested by clang-tidy

keneanung committed:
move clang-format file up, so tests are formatted properly as well

keneanung committed:
More formatting

SlySven commented:
@eowmob I am sorry but this one does seem to have gotten overlooked. Would it be possible for you to revisit this and merge in changes made to the rest of the Mudlet codebase, as fair as Qt versions go, Mudlet only supports Qt **5.14** as a minimum right now so that may mean some conditional compilation bits can be omitted - whether that applies to your submissions here I haven't fully checked. I did try to merge in the current codebase myself but since I am not *au fait* with MXP I could not tell which way some conflicts should be resolved in order to keep your proposal intact but consistent...

:slightly_frowning_face:

eowmob commented:
Hello Stephen,

 

well, I’m a bit busy now, with work in RL, family; and also some other hobbyist (widely mud related) coding. This is also why I did not answer quicker, sorry.

 

Still, I should be able to update my build environment (I don’t use that one for anything except Mudlet). For some time, I tried to keep track with the changes in the Mudlet main tree, but I found this very tedious, as it seemed the Github was kind of backing out other, generic changes to the mxp code, which I than had to reenter.. Well, it sounds as if you already worked a on reintegrating my MXP changes to the current tree. About conditional compilation bits, I can’t really remember having added or used any. But, I remember the Windows QT build environment I used (because that’s where I’d run Mudlet) lacked some ‘make’ features, so I had apply some tricks to build the test code.

 

Now, however, let’s be honest: Are you (and the Mudlet people) serious about this? See, I like standards as way to guarantee freedom to people, freedom to choose the client, freedom which game they play with their favourite OS or client. I happen to run a mud that allows to make quite extensive use of MXP.. but if you wish, you can still play it with ssh or even telnet (maybe lacking some colours or sound) – and this function will never be removed.

 

So, if a client comes across that claims and negotiates MXP compatibility, but then fails to display even simple things.. it makes me sad, and if possible I try to help fix it. Caused by a quite similar situation with mudportal, I actually ended up with a quite complete javascript / HTML5 interface to my mud (which I published on mudportal for people to use), where people don’t need any client (well, except a browser) to play the game.

 

But, my impression with Mudlet was that there is only little interest in fully supporting standards (admittedly, MXP was also defined by Zuggsoft for their client.. but there is a standards definition, and some other clients, for example mushclient, implement the standard even better than Zmud or Cmud).. anyway, for Mudlet, my impression was that the focus was on each mud having Mudlet specific macro packages and that only those shall allow for the full mud experience. So, I came to the conclusion that the Mudlet team sees no need in supporting MXP as a means to provide a better gaming experience to people w/o requiring the mud admin to provide a specialized mudlet profile.

 

So, again, I ask: Do you really intent to have better MXP support in the main source tree of Mudlet? If so, yes, I’d be willing to support you within my capabilities, even add stuff / try to fix things regarding MXP if I I’m not using/experiencing them in my own mud.

 

Looking forward to get some feedback from you about this,

Michael.

 

Von: Stephen Lyons ***@***.*** 
Gesendet: Donnerstag, 23. März 2023 11:46
An: Mudlet/Mudlet ***@***.***>
Cc: eowmob ***@***.***>; Mention ***@***.***>
Betreff: Re: [Mudlet/Mudlet] Improve: variety of MXP changes (#4200)

 

@eowmob <https://github.com/eowmob>  I am sorry but this one does seem to have gotten overlooked. Would it be possible for you to revisit this and merge in changes made to the rest of the Mudlet codebase, as fair as Qt versions go, Mudlet only supports Qt 5.14 as a minimum right now so that may mean some conditional compilation bits can be omitted - whether that applies to your submissions here I haven't fully checked. I did try to merge in the current codebase myself but since I am not au fait with MXP I could not tell which way some conflicts should be resolved in order to keep your proposal intact but consistent...

🙁

—
Reply to this email directly, view it on GitHub <https://github.com/Mudlet/Mudlet/pull/4200#issuecomment-1480966360> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARKJYG7ZO6CZT662GRHMAETW5QSX7ANCNFSM4S5SLGXQ> .
You are receiving this because you were mentioned.  <https://github.com/notifications/beacon/ARKJYGZKWTYB6GYQECZVSR3W5QSX7A5CNFSM4S5SLGX2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOLBC4BWA.gif> Message ID: ***@***.*** ***@***.***> >

vadi2 commented:
Hey -

> But, my impression with Mudlet was that there is only little interest in fully supporting standards (admittedly, MXP was also defined by Zuggsoft for their client.. but there is a standards definition, and some other clients, for example mushclient, implement the standard even better than Zmud or Cmud)

That is precisely the problem.

> for Mudlet, my impression was that the focus was on each mud having Mudlet specific macro packages and that only those shall allow for the full mud experience.

That's an opt-in option you can have, but we are keen on supporting MXP as much as we can. Not just MXP, but other standards like [MUD Client Media Protocol](https://wiki.mudlet.org/w/Standards:MUD_Client_Media_Protocol) and [MMP](https://wiki.mudlet.org/w/Standards:MMP) for example which standardize functionality across clients.

eowmob commented:
Vadim, Stephen:

 

Well as Github seemed to have various issues getting things together (especially with syncing changes third party modules) and you also mention using a new qt version. I decided to uninstall/remove my whole mudlet dir and the sdk, then made a clean clone of the development tree (and then try to get to the #4200 branch again, or maybe start a new one with my backups of the affected mxp source files).

 

Anyway using setup-windows-sdk.ps1 I can’t get the sdk to install.. PS version 5.1.x at some point it breaks complaining there is no C:\qt (indeed there is not), and at that start of the verbose log it says Using “C:\Qt\5.13.2\mingw73_32 as QT base directory.” What, at a first glance of the script that path seems to be some kind of default choice if no qt is detected.

 

Do I have to install qt or qtcreator first? Which version should I take?

 

Do you have any hints what to do?

 

Thx,

Michael.

 

Von: Vadim Peretokin ***@***.*** 
Gesendet: Samstag, 25. März 2023 10:10
An: Mudlet/Mudlet ***@***.***>
Cc: eowmob ***@***.***>; Mention ***@***.***>
Betreff: Re: [Mudlet/Mudlet] Improve: variety of MXP changes (#4200)

 

Hey -

But, my impression with Mudlet was that there is only little interest in fully supporting standards (admittedly, MXP was also defined by Zuggsoft for their client.. but there is a standards definition, and some other clients, for example mushclient, implement the standard even better than Zmud or Cmud)

That is precisely the problem.

for Mudlet, my impression was that the focus was on each mud having Mudlet specific macro packages and that only those shall allow for the full mud experience.

That's an opt-in option you can have, but we are keen on supporting MXP as much as we can. Not just MXP, but other standards like MUD Client Media Protocol <https://wiki.mudlet.org/w/Standards:MUD_Client_Media_Protocol>  and MMP <https://wiki.mudlet.org/w/Standards:MMP>  for example which standardize functionality across clients.

—
Reply to this email directly, view it on GitHub <https://github.com/Mudlet/Mudlet/pull/4200#issuecomment-1483772038> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARKJYGZVIEOIHR2QOBAACSTW52Y7XANCNFSM4S5SLGXQ> .
You are receiving this because you were mentioned.Message ID: ***@***.***>

vadi2 commented:
Try installing Qt 5.14.2, that is the one we are currently using.

eowmob commented:
Hello Vadim,

 

still no success setting up a new build environment, though I did a few
things.

 

setup-windows-sdk.ps1 does not start the Qt installing, basically it does
nothing at compiling and installing Qt\nDownloading

 

It just breaks right away complaining there is no \Qt directory.

 

I managed to find  qt-opensource-windows-x86-5.14.2.exe   but I was not sure
what modules to install. It offers zillions of mingw versions.

But have they been installed already, or not? Which version? Does it also
have the Qt libraries itself? Has a bunch of optional Qt libs.

 

Now, if I install this version B4 setup-windows-sdk.ps1 it is not detected, 

Still says Using C:\Qt\5.13.2\mingw73_32 as QT base directory.

 

Later on, it breaks, because there is no Qt\Tools. Indeed.. it is named
C:Qt\5.14.2\Tools

 

Somehow, it seems to me as if setup-windows-sdk.ps1 seems to refer to some
Qt versions no longer available for download and thus breaks,

but I was unable to find where this is defined.

 

Im not sure who of you is reading which channels, maybe we should move the
discussion about the build environment to the forum?

 

Michael.

 

Von: Vadim Peretokin ***@***.*** 
Gesendet: Sonntag, 26. März 2023 11:10
An: Mudlet/Mudlet ***@***.***>
Cc: eowmob ***@***.***>; Mention
***@***.***>
Betreff: Re: [Mudlet/Mudlet] Improve: variety of MXP changes (#4200)

 

Try installing Qt 5.14.2, that is the one we are currently using.


Reply to this email directly, view it on GitHub
<https://github.com/Mudlet/Mudlet/pull/4200#issuecomment-1484038948> , or
unsubscribe
<https://github.com/notifications/unsubscribe-auth/ARKJYG2GVTP5YNFQUQ6XISLW6
AIYNANCNFSM4S5SLGXQ> .
You are receiving this because you were mentioned.Message ID:
***@***.***>

SlySven commented:
Sorry to hear you are having so much bother getting a build environment set up. I must confess that, although I have done some development work in a Window environment it is a lot harder to set up compared to Linux. Indeed, with some recent changes in various packaged in the (not the main-line) MSYS2/Mingw-w64 setup I was using that has actually gotten broken on my main PC and so far I have not been able to get git (specifically `git-remote-https.exe` which has gotten borked by changes in `ssh` I believe) which means I cannot fetch other people's code from GitHub.

If you do happen to have access to/and are familiar with a Linux OS you may find things are a lot easier to do - and given that none of the MXP should contain anything that is OS specific any work in that environment will be equally valid to the Mudlet project. {Hell, I have even done some development from yet another OS, namely FreeBSD - and we also have contributors working/testing in MacOS}.

eowmob commented:
Hi,

 

Ive noticed now that the Qt package on the download site has been changed,
it is now called: qt-unified-windows-x64-online.exe (vs. x86).

Also, the installer does no longer support QT 5.14.2, only 5.15.2.   And it
somehow seems the libraries are only for mingw810. Or maybe only compiled
with 8.10? as you can still install 7.30. Im trying to find a working
combination of parameters, also tried modifying the install scripts. But
till now the parts dont fit together. Every now and then I start another
build attempt on my build computer  Seems now, the installer only installs
5.15.2 with mingw 8.10, but you can have 7.30 as an additional option too.

 

@Stephen: Yes, of course, for the MXP stuff the client OS should be
virtually pointless. Problem is, I dont have any linux installation with
develop system and/for GUI around. Only very small and simplistic server
systems for cli development. Id have to setup a new VM, large enough to
handle a QT IDE. Its a way, but until now it seemed easier to get the
windows IDE running.. well, well see.

 

Michael..

 

Von: Michael Weller ***@***.*** 
Gesendet: Montag, 27. März 2023 11:54
An: 'Mudlet/Mudlet'
***@***.***>;
'Mudlet/Mudlet' ***@***.***>
Cc: 'Mention' ***@***.***>
Betreff: AW: [Mudlet/Mudlet] Improve: variety of MXP changes (#4200)

 

Hello Vadim,

 

still no success setting up a new build environment, though I did a few
things.

 

setup-windows-sdk.ps1 does not start the Qt installing, basically it does
nothing at compiling and installing Qt\nDownloading

 

It just breaks right away complaining there is no \Qt directory.

 

I managed to find  qt-opensource-windows-x86-5.14.2.exe   but I was not sure
what modules to install. It offers zillions of mingw versions.

But have they been installed already, or not? Which version? Does it also
have the Qt libraries itself? Has a bunch of optional Qt libs.

 

Now, if I install this version B4 setup-windows-sdk.ps1 it is not detected, 

Still says Using C:\Qt\5.13.2\mingw73_32 as QT base directory.

 

Later on, it breaks, because there is no Qt\Tools. Indeed.. it is named
C:Qt\5.14.2\Tools

 

Somehow, it seems to me as if setup-windows-sdk.ps1 seems to refer to some
Qt versions no longer available for download and thus breaks,

but I was unable to find where this is defined.

 

Im not sure who of you is reading which channels, maybe we should move the
discussion about the build environment to the forum?

 

Michael.

 

Von: Vadim Peretokin ***@***.*** 
Gesendet: Sonntag, 26. März 2023 11:10
An: Mudlet/Mudlet ***@***.***>
Cc: eowmob ***@***.***>; Mention
***@***.***>
Betreff: Re: [Mudlet/Mudlet] Improve: variety of MXP changes (#4200)

 

Try installing Qt 5.14.2, that is the one we are currently using.


Reply to this email directly, view it on GitHub
<https://github.com/Mudlet/Mudlet/pull/4200#issuecomment-1484038948> , or
unsubscribe
<https://github.com/notifications/unsubscribe-auth/ARKJYG2GVTP5YNFQUQ6XISLW6
AIYNANCNFSM4S5SLGXQ> .
You are receiving this because you were mentioned.Message ID:
***@***.***>

eowmob commented:
Hi Vadim, hi Stephen,

So, as I now have a working build environment, what do you want me to do to get this in the main tree?

- Just resolve conflicts to make this compatible with the current development tree?

- Or, split it down into smaller bits. I guess that would mean to abandon the 4200 pull request, and made a few more. Issue here is, that there's a bunch of dependencies between the changes and actually, only if you have all of them, you'll get real MXP 1.0 compliance.  But well, if you insist, I could do it. But beware, again, there are dependencies and the missing bits of the current MXP implementation are not just a single line here and there. It will be difficult to cut this down into smaller pieces that just compile or even provide any seeable improvement.

- about testing, I did a bunch of it, I run a mud with MXP code everywhere and I have plenty of test code for corner stones of the MXP standard. Adding some test routines to Mudlet. Well, you asked me about it, and I fixed the existings tests, even added 1 or 2 more.. But all these tests are pretty pointless.

We are not talking about something like an sqrt library function where you can do 100000 random sqrt and then x*x to check the result. The MXP code deeply interacts with the terminal emulation and other parts of the code. Honestly, the only sensible test would be to have some text files with MXP and terminal codes embedded and then compare a screenshot of the result with an example. Anything else is pointless. Some change in the output routines can easily scramble the MXP code and the tests would not notice at all. Remember, all the tests were fine.. yet again, the whole entity/element MXP definitions in the main code tree do not work at all.

But again, what would you like me to do next?

Michael.

vadi2 commented:
> split it down into smaller bits. I guess that would mean to abandon the 4200 pull request, and made a few more

I think that would be the best, both for testability and fixing any issues that might arise.

> Honestly, the only sensible test would be to have some text files with MXP and terminal codes embedded and then compare a screenshot of the result with an example.

I have experience with this - and in fact we have it setup on https://www.rainforestqa.com already. One downside is that the platform last I checked compares in black and white.

I think traditional tests that we have for MXP can cover what we need, it's just harder to set them up. Mind having a look to see?

eowmob committed:
Merge remote-tracking branch 'upstream/development' into MXP-enhancements

eowmob committed:
Submodule changes commit

eowmob commented:
I played a bit around, trying to resync this branch to the original tree. But there were too many changes, it did not compile and all issues were in other, non mxp, modules. Anyway, you want it in smaller pieces, I'll open a new threat in the devel forum for this, revoke this pull request in a short while (I took notes of all comments here, to keep them in mind), and start with new branch doing much smaller steps.

vadi2 commented:
That works - thanks!
